### PR TITLE
Network Sharding

### PR DESCRIPTION
**Pull request by traian-schiau**

### CODE REVIEW

File: networksharding/kadSharder.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"math/big"
+)
+
+// KadGetShardID Callback returning the shard id of the given peer.ID
+type KadGetShardID func(peer.ID) uint32
+
+// KadSharder KAD based sharder
+//
+// Resets a number of MSb in the distance to nods sharding the same shard
+type KadSharder struct {
```
Comment (AdoAdoAdo) [inline networksharding/kadSharder.go]: do we need KadSharder exported if the fields are not exported?
we already have a constructor NewKadSharder

File: networksharding/kadSharder.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+func NewKadSharder(prioBits uint32, f KadGetShardID) (Sharder, error) {
+	if prioBits == 0 || f == nil {
+		return nil, fmt.Errorf("Bad params")
+	}
+	k := &KadSharder{
+		cutOffMask: 0,
+		gsFunc:     f,
+	}
+
+	if prioBits >= 8 {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder.go]: redone

File: networksharding/kadSharder_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
+package networksharding
+
+import (
+	"fmt"
+	"github.com/libp2p/go-libp2p-core/peer"
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/kadSharder_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/stretchr/testify/assert"
+	"math/big"
+	"testing"
+)
+
+func TestCutoOffBits(t *testing.T) {
+
+	i := []byte{0xff, 0xff}[:]
+
+	s, _ := NewKadSharder(1, fakeShard0)
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/kadSharder_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	k := s.(*KadSharder)
+	r := k.resetDistanceBits(i)
+	assert.Equal(t, big.NewInt(0).SetBytes(r), big.NewInt(0x7f<<8|0xff), "Shouls match")
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/kadSharder_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	r = k.resetDistanceBits(i)
+	assert.Equal(t, big.NewInt(0).SetBytes(r), big.NewInt(0xff), "Shouls match")
+
+	s, _ = NewKadSharder(9, fakeShard0)
+	k = s.(*KadSharder)
+	r = k.resetDistanceBits(i)
+	assert.Equal(t, big.NewInt(0).SetBytes(r), big.NewInt(0xff), "Shouls match")
+}
+
+func fakeShard0(id peer.ID) uint32 {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: kind of mock, but I prefer to keep it just in tests

File: networksharding/kadSharder_test.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	return 0
+}
+
+func fakeShardB2_1b(id peer.ID) uint32 {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: renamed

File: networksharding/kadSharder_test.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	s, _ := NewKadSharder(8, fakeShard0)
+	checkDistance(s, t)
+}
+
+func TestKadSharderOrdering2(t *testing.T) {
+	s, _ := NewKadSharder(2, fakeShardB2_1b)
+	checkOrdering(s, t)
+}
+
+const (
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/kadSharder_test.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		peerList[i] = peer.ID(fmt.Sprintf("NODE %d", i))
+	}
+	l1 := s.SortList(peerList, nodeA)
+
+	refShardID := fakeShardB2_1b(nodeA)
+	sameShardScore := uint64(0)
+	sameShardCount := uint64(0)
+	otherShardScore := uint64(0)
+	retLen := uint64(len(l1))
+	t.Logf("[ref] %s , sha %x, shard %d\n", string(nodeA), sha256.Sum256([]byte(nodeA)), refShardID)
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: I don't want to see this unless the test fails

File: networksharding/kadSharder_test.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	for i, id := range l1 {
+		// keep this
+		//sid := makeSortingID(id)
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/kadSharder_test.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		shardID := fakeShardB2_1b(id)
+
+		if shardID == refShardID {
+			sameShardScore += retLen - uint64(i)
+			sameShardCount++
+		} else {
+			otherShardScore += retLen - uint64(i)
+		}
+		// keep this
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/kadSharder_test.go]: done

File: networksharding/sharder.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+// Get the sharder
+func Get() Sharder {
+	return currentSharder
+}
+
+// Set the sharder, can only be done once
+func Set(s Sharder) error {
+	if s == nil {
+		return fmt.Errorf("The sharder cannot be nil")
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: done

File: networksharding/sharder.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	onceSetter.Do(func() {
+		currentSharder = s
+		ret = nil
+	})
+	return ret
+}
+
+// Sharder - Main sharder interface
+type Sharder interface {
+	// Get the shard id of the peer
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: done

[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: done

File: networksharding/sharder.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	GetShard(id peer.ID) uint32
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: to get the shard

File: networksharding/sharder.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+// Len is the number of elements in the collection.
+func (sl *sortingList) Len() int {
+	return len(sl.peers)
+}
+
+// Less reports whether the element with
+// index i should sort before the element with index j.
+func (sl *sortingList) Less(i int, j int) bool {
+	return sl.peers[i].distance.Cmp(sl.peers[j].distance) < 0
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: I expect this to be guaranteed by sort

File: networksharding/sharder.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// Swap swaps the elements with indexes i and j.
+func (sl *sortingList) Swap(i int, j int) {
+	sl.peers[i], sl.peers[j] = sl.peers[j], sl.peers[i]
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: I expect this to be guaranteed by sort

File: networksharding/sharder.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func keyFromID(id peer.ID) []byte {
+	key := sha256.Sum256([]byte(id))
+	return key[:]
+}
+
+func sortList(s Sharder, peers []peer.ID, ref peer.ID) []peer.ID {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: This is (should) the common sorting flow for all the sharding implementation

File: networksharding/noSharder.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
+package networksharding
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/noSharder.go]: done

File: networksharding/sharder.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+func (sl *sortingList) Less(i int, j int) bool {
+	return sl.peers[i].distance.Cmp(sl.peers[j].distance) < 0
+}
+
+// Swap swaps the elements with indexes i and j.
+func (sl *sortingList) Swap(i int, j int) {
+	sl.peers[i], sl.peers[j] = sl.peers[j], sl.peers[i]
+}
+
+func keyFromID(id peer.ID) []byte {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder.go]: i prefer to keep it like this

File: networksharding/sharder_test.go [hunk 20]
```diff
@@ -0,0 +1,77 @@
+package networksharding
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply networksharding/sharder_test.go]: done

File: p2p/libp2p/networksharding/errors.go [hunk 21]
```diff
@@ -0,0 +1,11 @@
+package networksharding
+
+import (
+	"errors"
+)
+
+var (
+	ErrAlreadySet = errors.New("already set")
```
Comment (AdoAdoAdo) [inline p2p/libp2p/networksharding/errors.go]: can you keep the same format as in the other errors.go file?

each exported error needs a comment

File: p2p/libp2p/networksharding/errors.go [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+	ErrBadParams  = errors.New("bad Pareams")
```
Comment (iulianpascalau) [inline p2p/libp2p/networksharding/errors.go]: Typo pareams. Maybe with lower case

File: p2p/libp2p/networksharding/kadSharder.go [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+type kadSharder struct {
+	prioBits  uint32
+	kGetShard KadGetShard
+}
+
+// NewkadSharder kadSharder constructor
+// prioBits - Number of reseted bits.
+// f - Callback used to get the shard id for a given peer.ID
+func NewKadSharder(prioBits uint32, kgs KadGetShard) (Sharder, error) {
+	if prioBits == 0 || kgs == nil {
```
Comment (iulianpascalau) [inline p2p/libp2p/networksharding/kadSharder.go]: Should check also on IsInterfaceNil

File: p2p/libp2p/networksharding/kadSharder.go [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+const (
+	maxMaskBits  = 8
+	fullMaskBits = 0xff
+)
+
+// KadGetShard peer to shard mapping interface
+type KadGetShard interface {
+	ByID(p2p.PeerID) uint32 //ByID get the shard id of the given peer.ID
```
Comment (iulianpascalau) [inline p2p/libp2p/networksharding/kadSharder.go]: Should add IsInterfaceNil

File: networksharding/kadSharder.go [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+// KadSharder KAD based sharder
+//
+// Resets a number of MSb in the distance to nods sharding the same shard
+type KadSharder struct {
+	cutOffMask byte
+	gsFunc     KadGetShardID
+}
+
+// GetShard Check networksharding.Sharder
+func (ks *KadSharder) GetShard(id peer.ID) uint32 {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply networksharding/kadSharder.go]: We should not have a generic sorter. This function should not be required

File: networksharding/sharder.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply networksharding/sharder.go]: I do not think it is required by the user struct of this interface. It should expose only sortlist function

File: networksharding/sharder.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply networksharding/sharder.go]: I would have embedded this in kadSharder implementation. In this way you do not need to expose the GetDistance and GetShard functions on the Sharder interface

File: networksharding/sharder.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply networksharding/sharder.go]: We should inject dependencies

File: p2p/libp2p/networksharding/errors.go (hunk 21 above)
Comment (traian-schiau) [reply p2p/libp2p/networksharding/errors.go]: done

File: p2p/libp2p/networksharding/errors.go (hunk 22 above)
Comment (traian-schiau) [reply p2p/libp2p/networksharding/errors.go]: done

File: p2p/libp2p/networksharding/kadSharder.go (hunk 24 above)
Comment (traian-schiau) [reply p2p/libp2p/networksharding/kadSharder.go]: done

File: p2p/libp2p/networksharding/kadSharder.go (hunk 23 above)
Comment (traian-schiau) [reply p2p/libp2p/networksharding/kadSharder.go]: done

File: p2p/libp2p/networksharding/kadSharder.go [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	maxMaskBits  = 8
+	fullMaskBits = 0xff
+)
+
+// KadGetShard peer to shard mapping interface
+type KadGetShard interface {
```
[the remark this answers is not in the document]
Comment (traian-schiau) [reply p2p/libp2p/networksharding/kadSharder.go]: done

**Pull request merged**